waku/waku_rln_relay/protocol_types.nim

```diff
@@ -67,54 +84,105 @@ type
+
+    let pb = initProtoBuffer(buffer)
+
+    var proof: seq[byte]
+    discard ? pb.getField(1, proof)
+    discard nsp.proof.copyFrom(proof)
+
+    var merkleRoot: seq[byte]
+    discard ? pb.getField(2, merkleRoot)
+    discard nsp.merkleRoot.copyFrom(merkleRoot)
+
+    var epoch: seq[byte]
+    discard ? pb.getField(3, epoch)
+    discard nsp.epoch.copyFrom(epoch)
+
+    var shareX: seq[byte]
+    discard ? pb.getField(4, shareX)
+    discard nsp.shareX.copyFrom(shareX)
 
-  var proof: seq[byte]
-  discard ? pb.getField(1, proof)
-  discard nsp.proof.copyFrom(proof)
+    var shareY: seq[byte]
+    discard ? pb.getField(5, shareY)
+    discard nsp.shareY.copyFrom(shareY)
 
-  var merkleRoot: seq[byte]
-  discard ? pb.getField(2, merkleRoot)
-  discard nsp.merkleRoot.copyFrom(merkleRoot)
+    var nullifier: seq[byte]
+    discard ? pb.getField(6, nullifier)
+    discard nsp.nullifier.copyFrom(nullifier)
 
-  var epoch: seq[byte]
-  discard ? pb.getField(3, epoch)
-  discard nsp.epoch.copyFrom(epoch)
+    var rlnIdentifier: seq[byte]
+    discard ? pb.getField(7, rlnIdentifier)
+    discard nsp.rlnIdentifier.copyFrom(rlnIdentifier)
 
-  var shareX: seq[byte]
-  discard ? pb.getField(4, shareX)
-  discard nsp.shareX.copyFrom(shareX)
+    return ok(nsp)
+else:
+  proc init*(T: type RateLimitProof, buffer: seq[byte]): ProtoResult[T] =
+    var nsp: RateLimitProof
+
+    let pb = initProtoBuffer(buffer)
+
+    var proof: seq[byte]
+    discard ? pb.getField(1, proof)
+    discard nsp.proof.copyFrom(proof)
 
-  var shareY: seq[byte]
-  discard ? pb.getField(5, shareY)
-  discard nsp.shareY.copyFrom(shareY)
+    var merkleRoot: seq[byte]
+    discard ? pb.getField(2, merkleRoot)
+    discard nsp.merkleRoot.copyFrom(merkleRoot)
 
-  var nullifier: seq[byte]
-  discard ? pb.getField(6, nullifier)
-  discard nsp.nullifier.copyFrom(nullifier)
+    var epoch: seq[byte]
+    discard ? pb.getField(3, epoch)
+    discard nsp.epoch.copyFrom(epoch)
 
-  var rlnIdentifier: seq[byte]
-  discard ? pb.getField(7, rlnIdentifier)
-  discard nsp.rlnIdentifier.copyFrom(rlnIdentifier)
+    var shareX: seq[byte]
+    discard ? pb.getField(4, shareX)
+    discard nsp.shareX.copyFrom(shareX)
 
-  return ok(nsp)
+    var shareY: seq[byte]
+    discard ? pb.getField(5, shareY)
+    discard nsp.shareY.copyFrom(shareY)
 
-proc encode*(nsp: RateLimitProof): ProtoBuffer =
-  var output = initProtoBuffer()
+    var nullifier: seq[byte]
+    discard ? pb.getField(6, nullifier)
+    discard nsp.nullifier.copyFrom(nullifier)
 
-  output.write3(1, nsp.proof)
-  output.write3(2, nsp.merkleRoot)
-  output.write3(3, nsp.epoch)
-  output.write3(4, nsp.shareX)
-  output.write3(5, nsp.shareY)
-  output.write3(6, nsp.nullifier)
-  output.write3(7, nsp.rlnIdentifier)
+    var rlnIdentifier: seq[byte]
+    discard ? pb.getField(7, rlnIdentifier)
+    discard nsp.rlnIdentifier.copyFrom(rlnIdentifier)
 
-  output.finish3()
+    return ok(nsp)
 
-  return output
+when defined(rln_v2):
```

Comment:
Would it be possible to only have one `encode` implementation if we only had `RateLimitProof`?

Reply:
addressed in 561a967e